### DISCUSSION TITLE
Link pybind11 module with a shared library that calls an external function

### DISCUSSION THREAD

**Discussion by shadowqwq** [pybind/pybind11#3356] [Q&A]
I'm trying to link a pybind11 module with a `.so` shared library, and the library calls functions not implemented in the `.so` file. It works fine in a normal c++ executable file, but raises `Undefined Symbol` error when imported in python.

Here is a simple demo to reproduce my problem.

The function `Student::print()` is complied to a dynamic library and it calls a function `Student::setId()` that is not included in the `.so` file. (If use `nm` command, it will show `U _ZN7Student5setIdEi`.)

It works fine in `main.cpp` to call `Student::print()`, but in `test.py`, it raises `ImportError: {mypath}/libstu_lib.so: undefined symbol: _ZN7Student5setIdEi`

To simplify, I set the external function to be its own member function. The problem still reproduces when calling functions belonging to another classes, so I think it doesn't matter.

Is there any option in pybind11 to deal with this problem? Because it is difficult to modify the source code of the dynamic library. Thanks.

**Student.h**
```
#include <iostream>
using namespace std;

class Student {
    public:
        Student(int id);
        void setId(int id);
        void print();
    private:
        int id;
};
```
**Student.cpp**
```
#include "Student.h"

Student::Student(int id) {
    this->id = id;
}

void Student::setId(int id) {
    this->id = id;
}
```
**Student_lib.cpp**
```
#include "Student.h"

void Student::print() {
    cout << "id: " << this->id << endl;
    this->setId(111);
    cout << "id: " << this->id << endl;
}
```
**Student_wrapper.cpp**
```
#include <pybind11/pybind11.h>
#include "Student.h"

namespace py = pybind11;

PYBIND11_MODULE(stu, m){
    py::class_<Student>(m, "Student")
        .def(py::init<int>())
        .def("setId", &Student::setId)
        .def("print", &Student::print);
}
```
**CMakeLists.txt**
```
cmake_minimum_required(VERSION 3.15)
project(test)

link_directories(${PROJECT_SOURCE_DIR})

add_library(stu_lib SHARED Student_lib.cpp)

add_executable(main main.cpp Student.cpp)
target_link_libraries(main stu_lib)

find_package(pybind11 REQUIRED)
pybind11_add_module(stu Student_wrapper.cpp Student.cpp)
target_link_libraries(stu PUBLIC stu_lib)
```
**main.cpp**
```
#include "Student.h"

int main(int argc, char** argv) {
    Student* s = new Student(1);
    s->print();
    return 0;
}
```
**test.py**
```
from stu import Student

s = Student(1)
s.print()
```

**Comment by nyue** [comment]
@shadowqwq Did you manage to find the reason or a solution ?

**Reply by shadowqwq** [reply]
Don't know how it works but there is a workaround: https://stackoverflow.com/a/69571412.

**Reply by SeniorCtrlPlayer** [reply]
Thanks, but i don't have source code of the shared library, how can i fix it?
